### PR TITLE
Change search to handle multiple resources and update search page

### PR DESCRIPTION
**Pull request by zherman0**
This updates the search page to transition to the new UX desgin. https://openshift.github.io/openshift-origin-design/admin-perspective/future-openshift/search/search
This will be a multi step process spread over version 4.4 and 4.5. This is step 1.
Instead of using patternfly4 select options, it uses an in house dropdown solution that includes a typeahead filter. (This is an alternative solution to PR3719, which is now closed)
Screenshot_2020-01-17 Search · OKD(1)
Screenshot_2020-01-17 Search · OKD(2)
Screenshot_2020-01-17 Search · OKD
Peek 2020-01-17 09-29

Since Events uses the same component, it too as changed.  The Events sections are now labelled to identify what is being shown and multiple resources can be selected.
Screenshot_2020-01-20 Events · OKD
Screenshot_2020-01-20 Events · OKD(1)
Screenshot_2020-01-20 Events · OKD(2)

### CODE REVIEW

**Comment by zherman0** [comment]
@spadgett @benjaminapetersen - Our home grown non-PF solution for the search dropdown. This is the third PR for search options.

**Comment by sg00dwin** [comment]
@zherman0 here's my branch with updates I mentioned I worked on yesterday. There's some conflicts with your most recent commit, which I think where addressing the same issue that I did. So I didn't open a pr against yours.

https://github.com/sg00dwin/console/commit/e9eb73ce4121783ac07e9822e5a91809533e639f

**gif showing how it looks**
multi-search

**Comment by zherman0** [comment]
There is an open issue with the search page that should be investigated with this work: https://issues.redhat.com/browse/CONSOLE-1981

**Comment by zherman0** [comment]
@spadgett @benjaminapetersen - This is ready for review.  While this was fairly large, the fact I changed search.jsx to search.tsx makes it a lot to look over.  Thanks.

File: frontend/public/components/search.tsx [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+      updateUrlParams('q', '');
+    }
+  };
+
+  const isKindExpanded = (kindView: string) => {
+    return collaspedState.includes(kindView);
+  };
+
+  const toggleIfKindExpanded = (kindView: string) => {
+    const collaspedItems: string[] = collaspedState.includes(kindView)
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: You should really be using a `Set` here. It makes the logic *much* simpler

File: frontend/public/components/search.tsx [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+      kind={kindObj.crd ? referenceForModel(kindObj) : kindObj.kind}
+      showTitle={false}
+      showTextFilter={false}
+      autoFocus={false}
+      mock={mock}
+    />
+  );
+});
+
+const updateUrlParams = (k, v) => {
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: Use the helpers in utils/router instead of writing your own

File: frontend/public/components/search.tsx [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+};
+const getResourceName = (v: string) => {
+  return v.split('~').pop();
+};
+
+const SearchPage_: React.FC<SearchProps> = (props) => {
+  const [selectedItems, setSelectedItems] = React.useState(['']);
+  const [collaspedState, setCollaspedState] = React.useState([]);
+  const [selector, setSelector] = React.useState(selectorFromString(''));
+  const [searchParameters, setSearchParameters] = React.useState<SearchParameters>({
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: You should have different state values for name and label. It's an antipattern to combine multiple values in an object with `useState`

File: frontend/public/components/search.tsx [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    return collaspedState.includes(kindView);
+  };
+
+  const toggleIfKindExpanded = (kindView: string) => {
+    const collaspedItems: string[] = collaspedState.includes(kindView)
+      ? collaspedState.filter((keepItem: string) => keepItem !== kindView)
+      : [...collaspedState, kindView as string];
+    setCollaspedState(collaspedItems);
+  };
+  const updateSearchFilter = (type: string, value: string) => {
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: I'd suggest separate helpers for updating name and for updating label

File: frontend/public/components/search.tsx [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+      <div className="co-search co-m-pane__body">
+        {selectedItems.map((item) => {
+          return (
+            <Expandable
+              key={item}
+              toggleText={
+                selectedItems.length === 1 || isKindExpanded(item)
+                  ? selectedItems.length === 1
+                    ? `${getResourceName(item)}`
+                    : `Hide ${getResourceName(item)}`
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: Avoid nested ternaries. They're hard to understand

File: frontend/public/components/search.tsx [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+      setSelector(selectorFromString(''));
+      updateUrlParams('q', '');
+    }
+  };
+
+  const isKindExpanded = (kindView: string) => {
+    return collaspedState.includes(kindView);
+  };
+
+  const toggleIfKindExpanded = (kindView: string) => {
```
Comment (spadgett) [inline frontend/public/components/search.tsx]:
```suggestion
  const toggleKindExpanded = (kindView: string) => {
```

File: frontend/public/components/resource-dropdown.tsx [hunk 7]
```diff
@@ -99,30 +99,36 @@ const ResourceListDropdown_: React.SFC<ResourceListDropdownProps> = (props) => {
   const kinds = resources.groupBy((m) => m.kind);
   const isDup = (kind) => kinds.get(kind).size > 1;
 
+  const isDropdownSelected = (kind: string) => {
```
[the remark this answers is not in the document]
Comment (spadgett) [reply frontend/public/components/resource-dropdown.tsx]: Or maybe `isKindSelected`

File: frontend/public/components/search-filter-dropdown.tsx [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    </DropdownItem>,
+    <DropdownItem key="name-action" name="Name" component="button">
+      Name
+    </DropdownItem>,
+  ];
+  const onValueChange = (value: string) => {
+    setInputValue(value);
+  };
+  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === 'Enter') {
```
[the remark this answers is not in the document]
Comment (zherman0) [reply frontend/public/components/search-filter-dropdown.tsx]: That's how the selector and the current app works today. Also, the UX design has no submit. So I just kept it as 'Enter' press.

File: frontend/public/components/search.tsx [hunk 9]
```diff
@@ -0,0 +1,284 @@
+import * as _ from 'lodash-es';
```
[the remark this answers is not in the document]
Comment (zherman0) [reply frontend/public/components/search.tsx]: jsx -> tsx will not compile without serious rework so in this case I don't think that will work

File: frontend/public/components/search.tsx [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+    let kind: string, q: string, name: string;
+
+    if (window.location.search) {
+      const sp = new URLSearchParams(window.location.search);
+      kind = sp.get('kind');
+      q = sp.get('q');
+      name = sp.get('name');
+    }
+    // Ensure that the "kind" route parameter is a valid resource kind ID
+    kind = kind ? decodeURIComponent(kind) : 'core~v1~Service';
```
[the remark this answers is not in the document]
Comment (zherman0) [reply frontend/public/components/search.tsx]: That decodeURIComponent is original code.  I will investigate if it is needed.

[the remark this answers is not in the document]
Comment (zherman0) [reply frontend/public/components/search.tsx]: I removed the default.

[the remark this answers is not in the document]
Comment (zherman0) [reply frontend/public/components/search.tsx]: on here, the decodeURI is not needed, however, it is needed on the label queries because of the = sign.

File: frontend/public/components/search.tsx [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+    setSelectedItems(kind.split(','));
+    const tags = split(_.isString(q) ? decodeURIComponent(q) : '');
```
[the remark this answers is not in the document]
Comment (zherman0) [reply frontend/public/components/search.tsx]: With decodeURI, the labels with form `app=something` do not properly get translated.  I will leave it here.

File: frontend/public/components/search.tsx [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+
+  const updateSelectedItems = (selection: string) => {
+    const updateItems: string[] = selectedItems.includes(selection as string)
+      ? selectedItems.filter((keepItem: string) => keepItem !== selection)
+      : [...selectedItems, selection as string];
+    setSelectedItems(updateItems);
+    updateUrlParams('kind', updateItems.map(encodeURIComponent).join(','));
+  };
+
+  const clearSelection = (v: string) => {
```
[the remark this answers is not in the document]
Comment (zherman0) [reply frontend/public/components/search.tsx]: I thought you might say that :)

File: frontend/public/components/search.tsx (hunk 2 above)
Comment (zherman0) [reply frontend/public/components/search.tsx]: That is original code again.  I will see if I can use a common util function

Comment (zherman0) [reply frontend/public/components/search.tsx]: This was original code but I have replaced it.

File: frontend/public/components/events.jsx [hunk 13]
```diff
@@ -136,31 +137,88 @@ export class EventsList extends React.Component {
...
     this.state = {
       type: 'all',
-      kind: 'all',
       textFilter: '',
+      selected: ['All'],
     };
   }
 
+  updateSelected = (selection) => {
+    let { selected } = this.state;
```
[the remark this answers is not in the document]
Comment (zherman0) [reply frontend/public/components/events.jsx]: Because the selected items are shown in the toolbar area as well and that area allows you to remove the items, I have to keep the selected items at the parent.

File: frontend/public/components/factory/list-page.jsx [hunk 14]
```diff
@@ -349,8 +352,10 @@ export const ListPage = withFallback((props) => {
     ListComponent,
     mock,
     name,
+    nameFilter,
```
[the remark this answers is not in the document]
Comment (zherman0) [reply frontend/public/components/factory/list-page.jsx]: I asked a question in regards to this a few days ago.  For some reason, the `name` does not propagate properly when I set it at the `AysncLoader` level so I needed to add a new prop.  I am sure there is a way to get it working.

File: frontend/public/components/search.tsx [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+      name: validNameTags,
+      label: validTags,
+    });
+  }, []);
+
+  const updateSelectedItems = (selection: string) => {
+    const updateItems: string[] = selectedItems.includes(selection as string)
+      ? selectedItems.filter((keepItem: string) => keepItem !== selection)
+      : [...selectedItems, selection as string];
+    setSelectedItems(updateItems);
```
[the remark this answers is not in the document]
Comment (zherman0) [reply frontend/public/components/search.tsx]: Same response as above, the parent has a toolbar which can modify the values so I have to keep control in the parent.

**Comment by zherman0** [comment]
@spadgett @benjaminapetersen - The code is ready for another review.  Thanks!

File: frontend/public/components/search.tsx [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+
+  const toggleKindExpanded = (kindView: string) => {
+    const newCollasped = new Set([...collaspedState]);
+    newCollasped.has(kindView) ? newCollasped.delete(kindView) : newCollasped.add(kindView);
+    setCollaspedState([...newCollasped]);
+  };
+
+  const updateNameFilter = (value: string) => {
+    setNameFilter([...nameFilter, value]);
+    setQueryArgument('name', [...nameFilter, value].map(encodeURIComponent).join(','));
```
Comment (spadgett) [inline frontend/public/components/search.tsx]:
```suggestion
    setQueryArgument('name', [...nameFilter, value].join(','));
```

File: frontend/public/components/search.tsx [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+  };
+
+  const updateLabelFilter = (value: string) => {
+    if (requirementFromString(value) !== undefined) {
+      setLabelFilter([...labelFilter, value]);
+      setSelector(selectorFromString([...labelFilter, value].join(',')));
+      setQueryArgument('q', [...labelFilter, value].map(encodeURIComponent).join(','));
```
Comment (spadgett) [inline frontend/public/components/search.tsx]:
```suggestion
      setQueryArgument('q', [...labelFilter, value].join(','));
```

File: frontend/public/components/search.tsx [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+  };
+
+  const updateSearchFilter = (type: string, value: string) => {
+    type === 'label' ? updateLabelFilter(value) : updateNameFilter(value);
+  };
+
+  const removeNameFilter = (value: string) => {
+    const newNames = nameFilter.filter((keepItem: string) => keepItem !== value);
+    setNameFilter(newNames);
+    setQueryArgument('name', newNames.map(encodeURIComponent).join(','));
```
Comment (spadgett) [inline frontend/public/components/search.tsx]:
```suggestion
    setQueryArgument('name', newNames.join(','));
```

File: frontend/public/components/search.tsx [hunk 19]
```diff
@@ -0,0 +1,264 @@
...
+  };
+
+  const removeLabelFilter = (value: string) => {
+    const newLabels = labelFilter.filter((keepItem: string) => keepItem !== value);
+    setLabelFilter(newLabels);
+    setQueryArgument('q', newLabels.map(encodeURIComponent).join(','));
```
Comment (spadgett) [inline frontend/public/components/search.tsx]:
```suggestion
    setQueryArgument('q', newLabels.join(','));
```

File: frontend/public/components/search.tsx [hunk 20]
```diff
@@ -0,0 +1,264 @@
...
+    setLabelFilter([]);
+    setNameFilter([]);
+    setSelector(selectorFromString(''));
+    setQueryArgument('kind', '');
+    setQueryArgument('q', '');
+    setQueryArgument('name', '');
+  };
+
+  const isKindExpanded = (kindView: string) => {
+    return collaspedState.includes(kindView);
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: It would be better to make this a `Set` as well

File: frontend/public/components/search.tsx [hunk 21]
```diff
@@ -0,0 +1,264 @@
...
+          />
+          <SearchFilterDropdown onChange={updateSearchFilter} />
+        </div>
+        <div className="form-group">
+          <ChipGroup withToolbar defaultIsOpen={false}>
+            <ChipGroupToolbarItem key="resources-category" categoryName="Resource">
+              {selectedItems.map((chip) => (
+                <Chip key={chip} onClick={() => updateSelectedItems(chip)}>
+                  <ResourceIcon kind={chip} />
+                  {chip.split('~').pop()}
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: Use `kindForReference`.

Note that this could get confusing if selecting the same kind name in different API groups. There'd be no way to tell them apart :/

File: frontend/public/components/search.tsx [hunk 22]
```diff
@@ -0,0 +1,264 @@
...
+                <span>
+                  <Button variant="plain" aria-label="Close" onClick={() => clearNameFilter()}>
+                    <CloseIcon />
+                  </Button>
+                </span>
+              )}
+            </ChipGroupToolbarItem>
+          </ChipGroup>
+          {(selectedItems.length > 0 || labelFilter.length > 0 || nameFilter.length > 0) && (
+            <Button variant="link" key="clear-filters" onClick={() => clearAll()}>
```
Comment (spadgett) [inline frontend/public/components/search.tsx]:
```suggestion
            <Button variant="link" key="clear-filters" onClick={clearAll}>
```

File: frontend/public/components/search.tsx [hunk 23]
```diff
@@ -0,0 +1,264 @@
...
+                mock={noProjectsAvailable}
+                key={item}
+              />
+            </Expandable>
+          );
+        })}
+        {selectedItems.length === 0 && (
+          <div className="text-center">
+            <br />
+            <br />
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: I'd avoid `br`. If you need more whitespace, can you use CSS?

We have existing styles for empty state than you should be able to use.

**Comment by sg00dwin** [comment]
@zherman0  I noticed a special case bug involving react virtualized scroll. It's happening when multiple resources are selected. When you scroll the top rows are hidden. Looking at the dom, the `ReactVirtualized__VirtualGrid__innerScrollContainer` is positioned correctly, but the table rows are positioned incorrectly and it appears some are hidden.  Here's an animated screenshot. 

multi-search-list-disappear2

File: frontend/public/components/search-filter-dropdown.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (zherman0) [reply frontend/public/components/search-filter-dropdown.tsx]: I agree this should be brought up with UX.  Is this an issue that we can shelve for this PR or do we need to resolve this first?

File: frontend/public/components/events.jsx [hunk 24]
```diff
@@ -20,12 +20,15 @@ import {
...
   pluralize,
+  ResourceIcon,
   ResourceLink,
   resourcePathFromModel,
   Timestamp,
   TogglePlay,
 } from './utils';
 import { EventStreamList } from './utils/event-stream';
+import { Button, Chip, ChipGroup, ChipGroupToolbarItem } from '@patternfly/react-core';
+import { CloseIcon } from '@patternfly/react-icons';
```
Comment (spadgett) [inline frontend/public/components/events.jsx]: Move these imports to the group at the top since they're not console imports

File: frontend/public/components/events.jsx [hunk 25]
```diff
@@ -136,43 +139,120 @@ export class EventsList extends React.Component {
...
               label="Events by name or message"
               onChange={(e) => this.setState({ textFilter: e.target.value || '' })}
             />
           </div>
-        </div>
-        <EventStream {...this.props} type={type} kind={kind} mock={mock} textFilter={textFilter} />
+          <div className="form-group">
+            <ChipGroup withToolbar defaultIsOpen={false}>
+              <ChipGroupToolbarItem key="resources-category" categoryName="Resource">
+                {[...selected].map((chip) => (
```
Comment (spadgett) [inline frontend/public/components/events.jsx]: Why not just...

```suggestion
                {selected.map((chip) => (
```

File: frontend/public/components/resource-dropdown.tsx [hunk 26]
```diff
@@ -43,9 +42,10 @@ const blacklistResources = ImmutableSet([
   'extensions/v1beta1.ReplicationControllerDummy',
 ]);
 
-const DropdownItem: React.SFC<DropdownItemProps> = ({ model, showGroup }) => (
+const DropdownItem: React.SFC<DropdownItemProps> = ({ model, showGroup, checked }) => (
   <>
-    <span className="co-resource-item">
+    <span className={'co-resource-item'}>
+      <Checkbox id={model.apiGroup + model.apiVersion + model.kind} isChecked={checked} />
```
Comment (spadgett) [inline frontend/public/components/resource-dropdown.tsx]: I'm not sure we'd see this in practice, but there's an edge case where this isn't unique.

group: foobar
version: baz
kind: qux

and

group: foo
version: barbaz
kind: qux

will give the same ID

I'd use `:` as a separator since it's allowed for `id` and isn't allowed for group/version/kind values

```suggestion
      <Checkbox id={`${model.apiGroup}:${model.apiVersion}:${model.kind}`} isChecked={checked} />
```

Is `id` really required here? I'd be best to just leave that off honestly unless Patternfly requires it.

File: frontend/public/components/events.jsx [hunk 27]
```diff
@@ -136,43 +139,120 @@ export class EventsList extends React.Component {
     super(props);
     this.state = {
       type: 'all',
-      kind: 'all',
       textFilter: '',
+      selected: new Set(['All']),
```
Comment (spadgett) [inline frontend/public/components/events.jsx]: nit: why `All` instead of `all`? I have a slight preference for the latter (and it's consistent with `all` for `type` above)

File: frontend/public/components/events.jsx [hunk 28]
```diff
@@ -136,43 +139,120 @@ export class EventsList extends React.Component {
...
+            <ChipGroup withToolbar defaultIsOpen={false}>
+              <ChipGroupToolbarItem key="resources-category" categoryName="Resource">
+                {[...selected].map((chip) => (
+                  <Chip key={chip} onClick={() => this.toggleSelected(chip)}>
+                    <ResourceIcon kind={chip} />
+                    {kindForReference(chip)}
+                  </Chip>
+                ))}
+                {selected.size > 0 && (
+                  <span>
```
Comment (spadgett) [inline frontend/public/components/events.jsx]: Is this span needed?

File: frontend/public/components/resource-dropdown.tsx [hunk 29]
```diff
@@ -99,30 +99,36 @@ const ResourceListDropdown_: React.SFC<ResourceListDropdownProps> = (props) => {
...
+      showGroup={isDup(model.kind)}
+      checked={isKindSelected(referenceForModel(model))}
+    />
   )) as OrderedMap<string, JSX.Element>;
-
   // Add an "All" item to the top if `showAll`.
   const allItems = (showAll
     ? OrderedMap({
-        all: (
+        All: (
```
Comment (spadgett) [inline frontend/public/components/resource-dropdown.tsx]: same question here: why the change to `All` from `all`? it doesn't look like this is the display value, just an internal string

File: frontend/public/components/resource-dropdown.tsx [hunk 30]
```diff
@@ -137,11 +143,17 @@ const ResourceListDropdown_: React.SFC<ResourceListDropdownProps> = (props) => {
       menuClassName="dropdown-menu--text-wrap"
       className={classNames('co-type-selector', className)}
       items={allItems}
-      title={allItems[selectedKey]}
+      title={
+        <div key={'title-resource'}>
```
Comment (spadgett) [inline frontend/public/components/resource-dropdown.tsx]: Is `key` needed? If it is, it should be

```suggestion
        <div key="title-resource">
```

But would be better to leave off if not required.

File: frontend/public/components/search-filter-dropdown.tsx [hunk 31]
```diff
@@ -0,0 +1,62 @@
+import * as React from 'react';
+import { Dropdown, DropdownToggle, DropdownItem, TextInput } from '@patternfly/react-core';
+import { CaretDownIcon, FilterIcon } from '@patternfly/react-icons';
+
+export const SearchFilterDropdown: React.SFC<SearchFilterDropdownProps> = (props) => {
+  const [isOpen, setOpen] = React.useState(false);
+  const [selected, setSelected] = React.useState('Label');
```
Comment (spadgett) [inline frontend/public/components/search-filter-dropdown.tsx]: Ideally we'd have an enum for label vs name instead of strings. If we use strings, use `label` (lowercase)

File: frontend/public/components/search-filter-dropdown.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (spadgett) [reply frontend/public/components/search-filter-dropdown.tsx]: > I agree this should be brought up with UX. Is this an issue that we can shelve for this PR or do we need to resolve this first?

Let's open a follow up issue.

File: frontend/public/components/search-filter-dropdown.tsx [hunk 32]
```diff
@@ -0,0 +1,62 @@
...
+                <FilterIcon className="span--icon__right-margin" /> {selected}
+              </>
+            </DropdownToggle>
+          }
+          isOpen={isOpen}
+          dropdownItems={dropdownItems}
+        />
+        <TextInput
+          onChange={setInputValue}
+          placeholder={selected === 'Label' ? 'Filter by label...' : 'Filter by name...'}
```
Comment (spadgett) [inline frontend/public/components/search-filter-dropdown.tsx]: We can't just use a plain text input for labels. We should use the `SelectorInput` component for labels

File: frontend/public/components/search.tsx [hunk 33]
```diff
@@ -0,0 +1,257 @@
...
+import { requirementFromString } from '../module/k8s/selector-requirement';
+import { ResourceListDropdown } from './resource-dropdown';
+import { resourceListPages } from './resource-pages';
+import { withStartGuide } from './start-guide';
+import { split, selectorFromString } from '../module/k8s/selector';
+import { referenceForModel, kindForReference } from '../module/k8s';
+import { LoadingBox, PageHeading, ResourceIcon } from './utils';
+import { Button, Chip, ChipGroup, ChipGroupToolbarItem, Expandable } from '@patternfly/react-core';
+import { SearchFilterDropdown } from './search-filter-dropdown';
+import { CloseIcon } from '@patternfly/react-icons';
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: Move Patternfly imports to the top group

File: frontend/public/components/search.tsx [hunk 34]
```diff
@@ -0,0 +1,257 @@
...
+import { connectToModel } from '../kinds';
+import { DefaultPage } from './default-resource';
+import { requirementFromString } from '../module/k8s/selector-requirement';
+import { ResourceListDropdown } from './resource-dropdown';
+import { resourceListPages } from './resource-pages';
+import { withStartGuide } from './start-guide';
+import { split, selectorFromString } from '../module/k8s/selector';
+import { referenceForModel, kindForReference } from '../module/k8s';
+import { LoadingBox, PageHeading, ResourceIcon } from './utils';
+import { Button, Chip, ChipGroup, ChipGroupToolbarItem, Expandable } from '@patternfly/react-core';
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: Move Patternfly imports to the top group

File: frontend/public/components/search.tsx [hunk 35]
```diff
@@ -0,0 +1,257 @@
...
+    const validNameTags = _.reject(nameTags, (tag) => requirementFromString(tag) === undefined);
+    setSelector(selectorFromString(validTags.join(',')));
+    setLabelFilter(validTags);
+    setNameFilter(validNameTags);
+  }, []);
+
+  const updateSelectedItems = (selection: string) => {
+    const updateItems = new Set([...selectedItems]);
+    updateItems.has(selection) ? updateItems.delete(selection) : updateItems.add(selection);
+    setSelectedItems([...updateItems]);
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: We can't we just store the `Set` in state?

File: frontend/public/components/search.tsx [hunk 36]
```diff
@@ -0,0 +1,257 @@
...
+      hideTextFilter
+      autoFocus={false}
+      mock={mock}
+    />
+  );
+});
+
+const SearchPage_: React.FC<SearchProps> = (props) => {
+  const [selectedItems, setSelectedItems] = React.useState([]);
+  const [collaspedState, setCollaspedState] = React.useState(new Set([]));
```
Comment (spadgett) [inline frontend/public/components/search.tsx]:
```suggestion
  const [collaspedState, setCollaspedState] = React.useState(new Set<string>([]));
```

File: frontend/public/components/search.tsx [hunk 37]
```diff
@@ -0,0 +1,257 @@
...
+  };
+
+  const clearAll = () => {
+    setSelectedItems([]);
+    setLabelFilter([]);
+    setNameFilter([]);
+    setSelector(selectorFromString(''));
+    setQueryArgument('kind', '');
+    setQueryArgument('q', '');
+    setQueryArgument('name', '');
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: I'd call your other helpers here instead of duplicating the logic.

```ts
clearSelectedItems();
clearNameFilter();
clearLabelFilter();
```

File: frontend/public/components/search.tsx [hunk 38]
```diff
@@ -0,0 +1,257 @@
...
+  };
+
+  const isKindExpanded = (kindView: string) => {
+    return collaspedState.has(kindView);
+  };
+
+  const toggleKindExpanded = (kindView: string) => {
+    const newCollasped = new Set(collaspedState);
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: We can't we just store the `Set` in state?

File: frontend/public/components/search.tsx [hunk 39]
```diff
@@ -0,0 +1,257 @@
...
+  const updateLabelFilter = (value: string) => {
+    if (requirementFromString(value) !== undefined) {
+      setLabelFilter([...labelFilter, value]);
+      setSelector(selectorFromString([...labelFilter, value].join(',')));
+      setQueryArgument('q', [...labelFilter, value].join(','));
+    }
+  };
+
+  const updateSearchFilter = (type: string, value: string) => {
+    type === 'label' ? updateLabelFilter(value) : updateNameFilter(value);
```
Comment (spadgett) [inline frontend/public/components/search.tsx]: We should use an enum for type

File: frontend/public/components/resource-dropdown.tsx (hunk 26 above)
Comment (zherman0) [reply frontend/public/components/resource-dropdown.tsx]: Id is required and it has to be unique.

File: frontend/public/components/events.jsx (hunk 27 above)
Comment (zherman0) [reply frontend/public/components/events.jsx]: No reason.  I will set it to `all`

Comment (zherman0) [reply frontend/public/components/events.jsx]: There is a reason, the resource icon expects "All" or else it shows "ALL" which is a change from prior.
image
I can add all kinds of code to capitalize the first letter, but seems messy.

File: frontend/public/components/resource-dropdown.tsx (hunk 29 above)
Comment (zherman0) [reply frontend/public/components/resource-dropdown.tsx]: I need it capitalized for simplicity.

File: frontend/public/components/resource-dropdown.tsx (hunk 30 above)
Comment (zherman0) [reply frontend/public/components/resource-dropdown.tsx]: Key is needed

File: frontend/public/components/search-filter-dropdown.tsx (hunk 31 above)
Comment (zherman0) [reply frontend/public/components/search-filter-dropdown.tsx]: Like the 'All' issue, I use the name programatically for the toolbar and I would have to capitalize the values otherwise.  For simplicity, I would like to leave this capitalized.

File: frontend/public/components/search-filter-dropdown.tsx (hunk 32 above)
Comment (zherman0) [reply frontend/public/components/search-filter-dropdown.tsx]: That's the PF TextInput.  So, the SelectorInput creates the tags that are in badges on the input field but I place those items in the new toolbar.  I need to keep this using the TextInput.

File: frontend/public/components/search.tsx (hunk 38 above)
Comment (zherman0) [reply frontend/public/components/search.tsx]: I am just making a copy there so I can then modify the state.  The collaspedState is now a Set.

File: frontend/public/components/events.jsx (hunk 25 above)
Comment (zherman0) [reply frontend/public/components/events.jsx]: map is not available on Sets, so i have to convert first to array

File: frontend/public/components/search.tsx (hunk 38 above)
Comment (spadgett) [reply frontend/public/components/search.tsx]: > I am just making a copy there so I can then modify the state. The collaspedState is now a Set.

OK, I see. We might make this an `ImmutableSet`, but this is fine.

File: frontend/public/components/search.tsx [hunk 40]
```diff
@@ -0,0 +1,253 @@
...
+  const [collaspedState, setCollaspedState] = React.useState(new Set<string>([]));
+  const [selector, setSelector] = React.useState(selectorFromString(''));
+  const [nameFilter, setNameFilter] = React.useState([]);
+  const [labelFilter, setLabelFilter] = React.useState([]);
+
+  const { namespace, noProjectsAvailable } = props;
+
+  // Set state variables from the URL
+  React.useEffect(() => {
+    let kind: string, q: string, name: string;
```
Comment (benjaminapetersen) [inline frontend/public/components/search.tsx]: You may be able to simplify/`const` some of this with destructuring and eliminate some conditional logic:

```typescript 
const {kind, q, name} = window.location.search 
    ? queryFromSearch() 
    : { kind: '', q: '', name: ''};
```

**Comment by benjaminapetersen** [comment]
Any interest in alphabetization, or should these list as you search?  When I added `CronJob` last, I didn't see it appear as I had `Deployment` and `StatefulSet` expanded. It works fine as is, but for a moment I wasn't sure if it picked up the change.

Screen Shot 2020-01-24 at 11 42 14 AM

File: frontend/public/components/factory/list-page.jsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (benjaminapetersen) [reply frontend/public/components/factory/list-page.jsx]: Yeah, this would be confusing later.

File: frontend/public/components/resource-dropdown.tsx [hunk 41]
```diff
@@ -137,11 +143,17 @@ const ResourceListDropdown_: React.SFC<ResourceListDropdownProps> = (props) => {
       menuClassName="dropdown-menu--text-wrap"
       className={classNames('co-type-selector', className)}
       items={allItems}
-      title={allItems[selectedKey]}
+      title={
+        <div key="title-resource">
+          Resources{' '}
+          <Badge isRead>
+            {selected.length === 1 && selected[0] === 'All' ? 'All' : selected.length}
```
Comment (benjaminapetersen) [inline frontend/public/components/resource-dropdown.tsx]: There is a lot of `All`, I'm a bit worried if we are depending on that for display.

**Comment by benjaminapetersen** [comment]
Experientally, I would prefer if it didn't automatically close on every checkbox select.  Its a bit cumbersome to select multiple items:



- [ ] Open bugzilla if we want to follow-up on this instead of fix on first pass.

**Comment by zherman0** [comment]
> Experientally, I would prefer if it didn't automatically close on every checkbox select. Its a bit cumbersome to select multiple items:
> 
> 
> 
>     * [ ]  Open bugzilla if we want to follow-up on this instead of fix on first pass.

@benjaminapetersen - I did discuss this with UX.  The thought was that most users would probably want only one item since that is how it works today.  If we force a user to close after one click then that greatly affects their experience. Now, if 80% of people do multiple search, then we should definitely leave it open.

**Comment by benjaminapetersen** [comment]
@zherman0 I would think "click outside" would close effectively, rather than require an explicit "close".

**Comment by benjaminapetersen** [comment]
It seems to be `&&` not `||`, requiring that a resource (or several) are selected before the label selector will work:

Screen Shot 2020-01-24 at 1 16 35 PM

I can imagine users wanting to search by label primarily, however.

- [ ] Decide if this is correct, or if we should fix/make a bug for follow.

File: frontend/public/components/resource-dropdown.tsx (hunk 41 above)
Comment (zherman0) [reply frontend/public/components/resource-dropdown.tsx]: I can switch that to an enum.

File: frontend/public/components/search.tsx (hunk 40 above)
Comment (zherman0) [reply frontend/public/components/search.tsx]: I like what you are doing here, but I would have to make the `queryFromSearch` function and that would be just as many lines as before and so we would not really save anything.

**Comment by spadgett** [comment]
> It seems to be `&&` not `||`, requiring that a resource (or several) are selected before the label selector will work:

There's no API to search across all resources by label. I'm not sure it's practical to run a separate list for every resource even with a label selector, so we ask the user to pick a resource first.

**Comment by benjaminapetersen** [comment]
Yeah thought about that after a bit, there is not `kubectl/oc` query with a label that doesn't require a `get x` resource.  So we don't have an API either.

**Comment by benjaminapetersen** [comment]
/lgtm
/hold

I think we are good to go with this one (if CI opens up). Can address any follows as bugs or future Jira if needed.

**Comment by benjaminapetersen** [comment]
Holding on #4065
